experimental/lib/Analysis/GsaAnalysis.cpp

```diff
@@ -0,0 +1,485 @@
+        expressionsList.emplace_back(phiInputCondition, operand);
+      }
+
+      // Sort the conditions according to their index
+      llvm::sort(conditionsList.begin(), conditionsList.end(),
+                 [](std::string a, std::string b) {
+                   a.erase(0, 1);
+                   b.erase(0, 1);
+                   return std::stoi(a) < std::stoi(b);
+                 });
+
+      // Expand the expressions to get the tree of gammas
+      auto *gammaRoot = expandGammaTree(expressionsList, conditionsList, phi);
+      gammaRoot->isRoot = true;
+
+      // Once that a phi has been converted into a tree of gammas, all the
+      // gates which used the original phi as input must be connected to the
+      // root of the gamma tree
+      for (auto &[bb, phis] : gateList) {
+        for (auto &phii : phis) {
+          for (auto &op : phii->operands) {
+            if (op->isTypeGate() && op->getGate() == phi)
+              op->input = gammaRoot;
+          }
+        }
+      }
+    }
+  }
+}
+
+template <typename FunctionType>
+void GsaAnalysis<FunctionType>::convertPhiToMu(FunctionType &funcOp) {
+
+  if (funcOp.getBlocks().size() == 1)
+    return;
+
+  mlir::DominanceInfo domInfo;
+  mlir::CFGLoopInfo loopInfo(domInfo.getDomTree(&funcOp.getBody()));
+
+  // For each phi
+  for (auto const &[phiBlock, phis] : gateList) {
+    for (auto *phi : phis) {
+
+      // A phi might be a MU iff it is inside a for loop and has exactly
+      // two operands
+      if (!loopInfo.getLoopFor(phiBlock) || phi->operands.size() != 2)
+        continue;
+
+      auto *op0Block = phi->operands[0]->getBlock(),
+           *op1Block = phi->operands[1]->getBlock();
+
+      // Checks whether the block of the merge is a loop header
+      bool isBlockHeader =
+          loopInfo.getLoopFor(phiBlock)->getHeader() == phiBlock;
+
+      // Checks whether the two operands come from different loops (in
+      // this case, one of the values is the initial definition)
+      bool operandFromOutsideLoop =
+          loopInfo.getLoopFor(op0Block) != loopInfo.getLoopFor(op1Block);
+
+      // If both the conditions hold, then we have a MU gate
+      if (!(isBlockHeader && operandFromOutsideLoop))
+        continue;
+
+      phi->gsaGateFunction = GateType::MuGate;
+
+      // Use the initial value of mu as first input of the gate
+      if (domInfo.dominates(op1Block, phiBlock))
+        std::swap(phi->operands[0], phi->operands[1]);
+
+      // The MU condition is given by the condition used for the termination
+      // of the loop
+      auto *terminator = loopInfo.getLoopFor(phi->getBlock())
+                             ->getExitingBlock()
+                             ->getTerminator();
+      phi->condition = getBlockCondition(terminator->getBlock());
+      phi->isRoot = true;
+    }
+  }
+}
+
+template <typename FunctionType>
+SmallVector<Gate *> *GsaAnalysis<FunctionType>::getGates(Block *bb) {
```

Comment:
This should return an `ArrayRef<Gate*>` if the returned vector should not be modifed. Here you are basically letting client code modify the internals of your class through the pointer.

Reply:
Should I find a way to have an array of `Gate`s rather than `Gate*` ?

Reply:
Yep missed that sorry. Problem is you cannot return an `ArrayRef` of stack-allocated `Gate`s because they will be out of scope when the function returns, and heap-allocating them seems overkill... I would say don't bother too much, and just return a `SmallVector<const Gate*>`, this is a larger object than `ArrayRef` but at least it has proper encapsulation.